### ISSUE TITLE
Incorrect truncation of logs

### ISSUE THREAD

**Issue by VivekThrivikraman-est** [rsyslog/rsyslog#5078]
### Expected behavior
Rsyslog should truncate large logs appropriately(not at every character) 

### Actual behavior
Looks like rsyslog is splitting messages at every byte, which seems to cause a flood of logs. Below are the logs which are coming if a large message is sent to rsyslog, looks like a new log is created(after truncation) after removing one character at a time, leading to huge number of logs. The below log is with imptcp, not sure if it would be the same for imtcp. This issue was not seen in 8.2106.0-150200.4.26.1 (with same config as given below).
Also looks like the message after truncation does not seem to be in syslog format.

> rsyslogd: imptcp imptcp: message received is at least **193** byte larger than max msg size; message will be split starting at: "189 <134>2023-02-03T08:04:02.720"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least **192** byte larger than max msg size; message will be split starting at: "89 <134>2023-02-03T08:04:02.7200"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least **191** byte larger than max msg size; message will be split starting at: "9 <134>2023-02-03T08:04:02.72006"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least **190** byte larger than max msg size; message will be split starting at: " <134>2023-02-03T08:04:02.720066"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 189 byte larger than max msg size; message will be split starting at: "<134>2023-02-03T08:04:02.720066+"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 188 byte larger than max msg size; message will be split starting at: "134>2023-02-03T08:04:02.720066+0"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 187 byte larger than max msg size; message will be split starting at: "34>2023-02-03T08:04:02.720066+01"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 186 byte larger than max msg size; message will be split starting at: "4>2023-02-03T08:04:02.720066+01:"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 185 byte larger than max msg size; message will be split starting at: ">2023-02-03T08:04:02.720066+01:0"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 184 byte larger than max msg size; message will be split starting at: "2023-02-03T08:04:02.720066+01:00"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 183 byte larger than max msg size; message will be split starting at: "023-02-03T08:04:02.720066+01:00 "  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 182 byte larger than max msg size; message will be split starting at: "23-02-03T08:04:02.720066+01:00 c"  [v8.2208.0]
> rsyslogd: imptcp imptcp: message received is at least 181 byte larger than max msg size; message will be split starting at: "3-02-03T08:04:02.720066+01:00 ci" 

### Steps to reproduce the behavior
Use the below config,use rsyslog version 8.2208.0, try sending logs bigger than 65536 bytes:

**config file 1:**

> input(
>      type="imptcp"
>      MaxFrameSize="65536"
>      SupportOctetCountedFraming="off"
>     port="30002"
>     ruleset="test"
> )
> 
> ruleset(name="test") {
>     set $!source = "test";
>     set $!softwareVersion = "9.6.1";
>     set $!sourceType = "xyz";
>     call forward
> }
> 

**parent config file:**

> 
> global(
>     net.enableDNS="off"
>     maxMessageSize="65536"
>     workDirectory="/work/da"
>     shutdown.enable.ctlc="on"
>     processInternalMessages="on"
>     DefaultNetstreamDriver="gtls"
>     DefaultNetstreamDriverCAFile="/config/certs/ca/rsyslog-ca.pem"
>     DefaultNetstreamDriverCertFile="/config/certs/server/rsyslog-server-cert.pem"
>     DefaultNetstreamDriverKeyFile="/config/certs/server/rsyslog-server-key.pem"
> )
> 
> main_queue(
>     queue.type="linkedlist"
>     queue.size="300000"
>     queue.dequeuebatchsize="10000"
>     queue.highwatermark="240000"
>     queue.lowwatermark="210000"
>     queue.filename="main_queue"
>     queue.maxdiskspace="500m"
>     queue.maxfilesize="100m"
>     queue.saveonshutdown="on"
>     queue.workerthreads="3"
> )
> 
> module(
>     load="imtcp"
>     MaxListeners="200"
>     MaxSessions="2000"
>     StreamDriver.Name="gtls"
>     StreamDriver.Mode="1"
>     StreamDriver.Authmode="x509/certvalid"
>     gnutlspriorityString="NORMAL:-VERS-TLS1.0:-VERS-TLS1.1"
> )
> 
> module(
>     load="impstats"
>     interval="60"
>     resetCounters="on"
>     ruleset="monitoring"
> )
> 
> module(load="imudp")
> module(load="imptcp")
> module(load="imrelp")
> module(load="omstdout")
> 
> template(name="testlogs" type="list" option.casesensitive="on") {
>     constant(value="{")
>     property(outname="DT" name="$!source" format="jsonf")
>     constant(value=", ")
>     property(outname="SV" name="$!softwareVersion" format="jsonf")
>     constant(value=", ")
>     property(outname="ST" name="$!sourceType" format="jsonf")
>     constant(value=", ")
>     constant(value="\"TIMESTAMP\":\"")
>     property(name="timegenerated" dateFormat="rfc3339" date.inUTC="on" position.from="1" position.to="23")
>     constant(value="Z\"")
>     constant(value=", ")
>     property(outname="HN" name="hostname" format="jsonf")
>     constant(value=", ")
>     property(outname="MESSAGE" name="rawmsg-after-pri" format="jsonf")
>     constant(value=", ")
>     property(outname="S" name="$!source" format="jsonf")
>     constant(value=", ")
>     property(outname="FAC" name="syslogfacility" format="jsonf")
>     constant(value=", ")
>     property(outname="SEV" name="syslogseverity" format="jsonf")
>     constant(value=", ")
>     property(outname="APPNAME" name="app-name" format="jsonf")
>     constant(value="}")
>     constant(value="\n")
> }
> 
>  template(name="container-out" type="string" string="%timegenerated:::date-rfc3339% [%syslogseverity-text%]: %msg%")
> action(name="container-out" type="omstdout" template="container-out")
> 
>  include(
>     file="/config/imtcp.conf"
>     mode="optional"
> )
> 
> include(
>     file="/work/services.d/*.conf"
>     mode="optional"
> )
> 
> input(
>     type="imptcp"
>     port="10514"
>     ruleset="availability_check"
> )
> ruleset(name="availability_check") { stop }
> 
> ruleset(name="forward") {
>     action(
>         name="forward_to_pipeline"
>         type="omfwd"
>         target="127.0.0.1"
>         port="6000"
>         protocol="tcp"
>         template="testlogs"
>         action.resumeretrycount="-1"
>         action.resumeinterval="1"
>         queue.type="linkedlist"
>         queue.size="300000"
>         queue.dequeuebatchsize="10000"
>         queue.highwatermark="240000"
>         queue.lowwatermark="210000"
>         queue.filename="forward_queue"
>         queue.maxdiskspace="500m"
>         queue.maxfilesize="100m"
>         queue.saveonshutdown="on"
>         queue.timeoutshutdown="1000"
>         queue.workerthreads="3"
>     )
> }
> 
> ruleset(name="monitoring") {
>     action(
>         name="impstats_to_logkf"
>         type="omfwd"
>         target="127.0.0.1"
>         port="6001"
>         protocol="tcp"
>         action.resumeretrycount="-1"
>         action.resumeinterval="1"
>         queue.type="FixedArray"
>         queue.highwatermark="240000"
>         queue.spoolDirectory="/var/run/rsyslog/queues"
>         queue.filename="stats_ruleset"
>         queue.lowwatermark="180000"
>         queue.maxdiskspace="300m"
>         queue.size="300000"
>         queue.dequeuebatchsize="10000"
>         queue.saveonshutdown="on"
>     )
> }
> 

### Environment
- rsyslog version: 8.2208.0
- platform: SUSE Linux Enterprise Server 15 SP4
- for configuration questions/issues, include rsyslog.conf and included config files : Included above

**Comment by davidelang** [comment]
by definition, splitting a log at any point is going to result in the bytes 
after that point not being in the proper syslog format.

what do you consider 'appropriate' truncation of logs that exceed the max length 
that the admin has configured? (since you don't think that truncating at that 
length is appropriate?


Rsyslog has an option to either throw away the remainder of the log when it 
hits the maxmessagesize (looking for the next newline to find the start of the next message) or to 
store the remainder as a separate log message (recognizing that it's going to be 
malformed)

In your case, it looks like you don't have a proper newline at the end of the 
message, so rsyslog has no way of knowing when the next log message should 
start.

A human can look at this and say "it obviously should start with the <134> as 
that is part of a properly formatted message" but this sort of logic leap is not 
what computers do well (and would break cases where that sort of thing appears 
inside a message)

David Lang

**Comment by VivekThrivikraman-est** [comment]
HI @davidelang 

Thanks for the response.I agree to what you have put down, my problem is slightly different here. If the message is bigger than the max message size, why does it have to split after every character which exceeds, and this happens even if there is a newline at the end. The behavior that I am seeing is: consider the string "123456789" to be the part exceeding max size, why should rsyslog create messages like "23456789", "3456789","456789" ... etc ? which is creating flood of messages and eventually rsyslog service dies(this is what I could understand from the logs that are pasted in the description). I have attached a new set of log which is with newline at the end(rsys_largemessage.log). 
What I thought should have happened is, if say max size is say 10kb and message of size say 22kb comes, then there should be 3 log messages max 10 + 10 +2 sizes respectively if split is enabled(and not truncation).

Also, this behavior was not seen in 8.2106.0-150200.4.26.1, only after upgrading to release 8.2208.0 we are seeing this(in older release with same config, the truncation of the message used to happen(no split) and rsyslog service was stable).

Can you also please let me know what is the proper way to configure truncation of messages(imtcp and imptcp)? I am trying out with the below config changes, I still see some split(thought I have set truncate as mode) happening though not as bad as the previous config, basically there is no flooding with it, but I still see some split happening, and I am yet to checkout what and where it is splitting:

In global config added oversizemsg.input.mode="truncate":
          global(
    net.enableDNS="off"
    maxMessageSize="65536"
	**oversizemsg.input.mode="truncate"**
In imptcp input added DiscardTruncatedMsg="on": 
      input(
               type="imptcp"
              MaxFrameSize="65536"
             **DiscardTruncatedMsg="on"** 
And in imtcp added DiscardTruncatedMsg="on":
    module(
        load="imtcp"
        **DiscardTruncatedMsg="on"**
      
[rsys_largemessage.log](https://github.com/rsyslog/rsyslog/files/10660579/rsys_largemessage.log)


Thanks,
Vivek

**Comment by VivekThrivikraman-est** [comment]
An update, with the configuration I pasted in the previous comment (ie using DiscardTruncatedMsg="on" for imptcp and oversizemsg.input.mode="truncate" in global) the flooding of message is no more happening, but I see another issue wherein the message after applying template(used in omfwd,which creates a json) is truncated and the rest of the response(after truncation,ie the second part) has proper json(template properly applied).

For eg: Say the raw message is 123456789, and say this message exceeds the maxMessageSize, what I am seeing in tcpdump as output is:
{DT:"test",SV:"1","ST":"X","TIMESTAMP":"2023-02-09T14:59:51.922","HN","h1","MESSAGE":"**123456**
{DT:"test",SV:"1","ST":"X","TIMESTAMP":"2023-02-09T14:59:51.922","HN","h1","MESSAGE":"789"}

Two issues with the above output:
1) The first line as you see has not applied template completely, resulting in incorrect json(I think it should have truncated the rawmsg and then the template should have been applied, but looks like truncation is done after applying template).
2)With DiscardTruncatedMsg="on", I can still see split happening.

I am attaching the actual tcpdump output, gave example above as the logs are big.
[tcpdump.txt](https://github.com/rsyslog/rsyslog/files/10698699/tcpdump.txt)

**Comment by rgerhards** [comment]
sry for the late reply. This root cause is a regression in commit f052717178. It was not really obvious and the testbench test did not detect this error case. I'll craft a fix for it.

Regarding where the message is truncated if it is truncated: this is an error measure, so we can not look into the semantics of the message. The point where this issue happens is when we actually only have a string of characters without any semantics.
